Player Vending Machines

penny64 opened the pull request:
<!-- The text between the arrows are comments - they will not be visible on your PR. -->
<!-- To automatically tag this PR, add the uppercase label(s) surrounded by brackets below, for example: [LABEL] -->
[FEATURE]

## About the PR <!-- Describe the Pull Request here. What does it change? What other things could this impact? -->

This adds constructible custom vending machines to the game.

It does this by adding vending machine frames to the steel construction menu and providing "vending machine modules" based on restock modules, which are installed in these frames to build vending machines.

![image](https://user-images.githubusercontent.com/13049028/115098902-8f13b700-9ee7-11eb-81c8-e4a63f44e3dd.png)

A "vending machine module" is found in the general manufacturer, which is used to make player vending machines.

![image](https://user-images.githubusercontent.com/13049028/115098881-6f7c8e80-9ee7-11eb-8305-283b6f5f7a6c.png)

An unused Val-U-Chimp vending module is defined currently as an example, it could be used in the future to replace the monkey crate in QM.

![image](https://user-images.githubusercontent.com/13049028/115098936-b4a0c080-9ee7-11eb-9e1d-cccaac3f8611.png)

Players can register their ID with it to receive payments directly to their bank account, add items, and set the prices and the name of the vendor! 
The registered owner's name is readily available by looking in the service panel, allowing security some chance of catching the people creating C-saber vendors.

To build them:

- Construct a vending machine frame from 3 steel sheets
- Wrench down the machine
- Install a vending module- found in general manufacturers
- Install 5 wires
- Install 2 glass
- screwdriver it to activate

Then to operate:

- Scan your ID card and enter your PIN
- Open the service panel and click on "Unregistered (locked)" to register the machine
- Load items into it while the chute is open
- Select an icon to promote on the screen with (*)
- Set prices
- Click your name to lock the machine, close the service panel, and enjoy payments directly to your bank account! 

For clear balance reasons, NT has fixed the "throwing inventory at users" bug that used to occur in older models when certain electrical interference is received.

Messaging for bank account payments is planned but will be in a future PR along with a selection of sprites you can display on the screen! This PR is everything I want to do in one PR, it's completely fleshed out.

This model has also received a new "hi-def" full color screen as the promotional display to help sell it's delightful array of inventory!
This model offers a generous cut to NT employees with a mere 10% to your local Quartermasters! Just another reason that you signed the contract.

![image](https://user-images.githubusercontent.com/13049028/113965651-a44f5e00-97e2-11eb-81e7-acc5775d5b55.png)

![image](https://user-images.githubusercontent.com/13049028/115094053-10138400-9ed1-11eb-97f6-07b6d3a71fcb.png)

![image](https://user-images.githubusercontent.com/13049028/115096643-e65f5a80-9eda-11eb-9a4d-27c5f482f2e8.png)

![Player vending machine](https://cdn.discordapp.com/attachments/814938775624351746/827398284984844298/youvend.gif)
Sprite credit Ludolf


## Why's this needed? <!-- Describe why you think this should be added to the game. -->

Goon isn't setup for an economy and this won't make it have one, but having this small piece of a player economy will be extremely fun all across departments- cargo can use it as a source of funds, science can sell beneficial chemicals (or not so beneficial), HOP may force players who lose their ID to buy a blank one- and more!
Not to mention the mischief potential.
I understand this idea really sells itself, I've seen tons of requests from this and optimism about it from everyone. It's fun.
## Changelog
<!-- If necessary, put your changelog entry below. Otherwise, please delete it.
Use however you want to be credited in the changelog in place of CodeDude.
Use (*) for major changes and (+) for minor changes. For example: -->

```changelog
(u)Penny & Ludolf
(*)Players can now construct vending machines to sell their own items. Find the frames in the steel sheet menu and vending modules in your local general manufacturer! Sprites and animations by Ludolf!
```

mathis650 commented:
poggers

ZeWaka commented on code/modules/vending/vending.dm:
```diff
@@ -1587,6 +1602,50 @@
+//And thus you will be burdened
+/obj/machinery/vending/player
+	name = "YouVend"
+	icon_state = "player"
```
hey so like i know this is somewhat out of scope, but it'd be cool if we could like change the appearance a bit

idea:
not sure on the actual mechanics, but hitting the machine with an item (maybe like drag drop? not sure.) would add an overlay of the item (50% scale, would be distorted unlike this) onto the machine
![image](https://user-images.githubusercontent.com/4741640/113760724-6e6b8600-96cb-11eb-81a4-f005c17fdba7.png)

Azrun replied:
Or what if it cycled through vendable objects?

penny64 replied:
I think Azrun's ideal is perfect, but I think it's out of scope for the initial PR! I like to compartmentalize my changes- not to mention get merged sooner :)

penny64 replied:
You can now choose an item in the service panel as the icon

penny64 replied:
You can now select products in the service menu as the promo icon

penny64 commented:
I'm closing in on these
What's left:
Group labeled Items
Bank account payments
Display item icons on screen
Set item price
Fallen over/fallen over broken sprite

Laboredih123 commented:
we need this

penny64 commented:
All that's left is some sprite work gonna open this up to review

penny64 commented:
I'm only missing a sprite for the module itself, these are done!

penny64 commented:
(modules sprited btw)

Azrun commented on code/modules/vending/vending.dm:
```diff
@@ -1586,7 +1616,335 @@
+		else if(href_list["vend"] && (length(player_list) <= 0))
+			src.setItemOverlay(null)
+			icon_state = "player"
+		else if(href_list["icon"] && src.panel_open && src.unlocked)
+			var/datum/data/vending_product/player_product/R = locate(href_list["icon"]) in src.player_list
+			src.setItemOverlay(R.icon)
+			src.generate_HTML(1, 0)
```
![image](https://user-images.githubusercontent.com/1017374/115107762-b6877580-9f21-11eb-9362-84860b6c3af0.png)
Overlays seem to get wonky.... Can only seem to update image if I stick something in and only to the thing I put in?

![image](https://user-images.githubusercontent.com/1017374/115108420-38c56900-9f25-11eb-9222-a7ad5ffb701c.png)
![image](https://user-images.githubusercontent.com/1017374/115108491-9e195a00-9f25-11eb-9768-9b825069eced.png)


Consider a general function to update all the overlays to things don't get out of sync.

penny64 replied:
>Overlays seem to get wonky.... Can only seem to update image if I stick something in and only to the thing I put in?

This is the only functionality available in this PR, in the future you'll be able to choose from a selection of promotional images.

Azrun replied:
Then only show anchor astrix for items that will work, otherwise it feels broken.

Unable to clear loading chute, always depicted.

penny64 replied:
Completely impossible, the asterisk is there so you can choose ones that work.
I don't know anything about product sprites, I have 32x32 and I shrink them to a size that lets the widest items fit.
This is all that can be done.

penny64 replied:
I'll resolve this cause anything else will be about the manager

Azrun commented on code/modules/vending/vending.dm:
```diff
@@ -1586,7 +1616,335 @@
+				loading = !loading
+				src.generate_HTML(0, 1)
+		else if (href_list["unlock"] && src.panel_open)
+			if (!owner && src.scan?.registered)
+				owneraccount = FindBankAccountByName(src.scan.registered)
+				owner = src.scan.registered
+				cardname = src.scan.name
+				unlocked = 1
+				loading = 1
```
Could it just register to whoever swipes first? Should you be able to... re-register. 😈

penny64 replied:
Unregistering is impossible so the cops can catch you

Azrun replied:
~Emag to clear registered owner?~

penny64 replied:
It randomizez the owner

Azrun commented on code/modules/vending/vending.dm:
```diff
@@ -1586,7 +1616,335 @@
+		else if (href_list["setprice"] && src.panel_open && src.unlocked)
+			var/inp
+			inp = input(usr,"Enter the new price:","Item Price", "") as num
+			if(inp && (usr.stat || usr.restrained() || in_interact_range(src, usr)))
```
Validate player input negative values take ~10% from shipping....

penny64 commented on code/modules/vending/vending.dm:
```diff
@@ -338,7 +339,17 @@
 				html_parts += "<tr><td>[R.product_name]</a></td><td colspan='2' style='text-align: center;'><strong>SOLD OUT</strong></td></tr>"
-
+		if(player_list)
+			var/obj/machinery/vending/player/T = src
+			for (var/datum/data/vending_product/player_product/R in src.player_list)
+				var/obj/item/productholder = R.contents[1]
+				var/nextproduct = html_encode(sanitize(productholder.name))
+				if (!T.unlocked == 1)
+					html_parts += "<tr><td><a href='byond://?src=\ref[src];vend=\ref[R]'>[nextproduct]</a></td><td style='text-align: right;'>[R.product_amount]</td><td style='text-align: right;'> $[R.product_cost]</td></tr>"
+					//Player vending machines don't have "out of stock" items
+				else if (!T.unlocked == 0)
+					//Links for setting prices when player vending machines are unlocked
+					html_parts += "<tr><td><a href='byond://?src=\ref[src];vend=\ref[R]'>[nextproduct]</a></td><td style='text-align: right;'>[R.product_amount]</td><td style='text-align: right;'><a href='byond://?src=\ref[src];setprice=\ref[R]'>$[R.product_cost]</a> (<a href='byond://?src=\ref[src];icon=\ref[R]'>*</a>)</td></tr>"
```
No there's no way to extend static HTML code a small selection of changes were made in the parent to accommodate player machines

penny64 commented on code/modules/vending/vending.dm:
```diff
@@ -1586,7 +1616,335 @@
+	attackby(obj/item/target, mob/user)
+		if(iswrenchingtool(target))
+			playsound(src.loc, "sound/items/Ratchet.ogg", 50, 1)
+			if(!wrenched && do_after(user, 2 SECONDS))
```
Actionbar in PR 2 no more features

Azrun replied:
I was under the impression generic actionbar was intended to replace do_after(), do_after() isn't a very satisfying UX. 🤷

ZeWaka replied:
Yep generic would be perfect here

penny64 replied:
Actionbar done

penny64 commented on code/modules/vending/vending.dm:
```diff
@@ -338,7 +339,17 @@
+			for (var/datum/data/vending_product/player_product/R in src.player_list)
+				var/obj/item/productholder = R.contents[1]
+				var/nextproduct = html_encode(sanitize(productholder.name))
+				if (!T.unlocked == 1)
```
:/ the =='s?

Azrun replied:
Correct not very readable.

penny64 replied:
Okay fixed!

penny64 commented on code/modules/vending/vending.dm:
```diff
@@ -593,8 +607,21 @@
+				else
+					//Players get 90% of profit from player vending machines QMs get 10%
+					var/obj/machinery/vending/player/T = src
+					T.owneraccount.fields["current_money"] += round(R.product_cost * profit)
+					wagesystem.shipping_budget += round(R.product_cost / (profit * 10))
```
I think money leaving is kind of a system level thing, and these are kind of a personal player thing. Either I give QMs less money- when I'm barely sending them anything, or I do enough fees to actually make it annoying for players.
I tested a lot and this feels right.

penny64 commented on code/modules/vending/vending.dm:
```diff
@@ -1586,7 +1616,335 @@
+				boardinstalled = 1
+		else if(istype(target, /obj/item/cable_coil))
+			var/obj/item/cable_coil/targetcoil = target
+			playsound(src.loc, "sound/items/Deconstruct.ogg", 50, 1)
+			if(targetcoil.amount >= 5 && boardinstalled && !wiresinstalled && do_after(user, 2 SECONDS))
```
I like playing the sound before the do_after, the other stuff was really inconsistent about it but this is what felt good

Azrun replied:
If board isn't installed, wires are already installed, or coils < 5 it will play the deconstruct sound every time you click it, that is how I found it.

penny64 replied:
Ah! You're right!

penny64 replied:
This is fixed we use an actionbar now so the sound just playes after

penny64 commented on code/modules/vending/vending.dm:
```diff
@@ -1586,7 +1616,335 @@
+				var/datum/data/vending_product/player_product/R = locate(href_list["setprice"]) in src.player_list
+				R.product_cost = inp
+				src.generate_HTML(1, 0)
+		else if(href_list["vend"] && (length(player_list) <= 0))
```
I'm not sure what you mean to check if it's "still src.player_list", additionally, we actually delete any empty item entries. And we don't display them, if somehow there was one with 0

Azrun replied:
If you vend a "demo" item with a quantity of one item which will be removed from the player_list, is the intent the machine displays an unavailable product when there are other products to display.

penny64 replied:
Oh yeah, I don't want to adjust that.
Intent is you can use a promo image of anything you get in there and keep it, if you like. 
I want some machines to have stupid promo images on them but not a onslaught

Azrun replied:
🤷 👍

penny64 commented on code/modules/vending/vending.dm:
```diff
@@ -1586,7 +1616,335 @@
+				boutput(user, "<span class='notice'>You add cables to the frame.</span>")
+			else if(!wiresinstalled && boardinstalled)
+				boutput(user, "<span class='alert'>You need at least five pieces of cable to wire the vending machine.</span>")
+		else if(istype(target, /obj/item/sheet) && wiresinstalled && !glassed)
+			var/obj/item/sheet/S = target
+			if (!(S.material && S.amount >= 2 && S.material.material_flags & MATERIAL_CRYSTAL))
```
I dunno I'd like to take any kind of glass, I think I'll make it eat the glass you give it and give it back when you pry it out.

Azrun replied:
Consider subdividing the stack and shoving into the frame's contents, that way what goes in comes out or have it only accept glass.

It is neat you can use koshmarite or the like, though that material doesn't seem very translucent...

penny64 replied:
You're right I'll just take glass, this kind of silliness is old goon

penny64 replied:
And done

Azrun commented on code/modules/vending/vending.dm:
```diff
@@ -1587,6 +1618,362 @@
+			user.visible_message("<b>[user.name]</b> loads [target] into [src].")
+			return
+		var/action = input(user, "What do you want to do with [targetContainer]?") as null|anything in list("Empty it into the vending machine","Place it in the vending machine")
+		if(action == "Place it in the vending machine" && (user.stat || user.restrained() || in_interact_range(src, user)))
```
Unconscious, Dead or in range? 😕

penny64 replied:
heh

ZeWaka commented on code/datums/manufacturing.dm:
```diff
@@ -599,10 +599,10 @@ proc/get_nice_mat_name_for_manufacturers(mat)
 ////////////////////////////////
 
 /datum/manufacture/circuit_board
```
^

ZeWaka commented on code/modules/vending/vending.dm:
```diff
@@ -1587,6 +1618,385 @@
+			user.visible_message("<b>[user.name]</b> loads [target] into [src].")
+			return
+		var/action = input(user, "What do you want to do with [targetContainer]?") as null|anything in list("Empty it into the vending machine","Place it in the vending machine")
+		if(action == "Place it in the vending machine" && !(user.stat || user.restrained() || !in_interact_range(src, user)))
```
don't use `mob.stat` directly, macros please
also, check for incapacitation not just restrained, there is a macro for this

ZeWaka commented on code/modules/vending/vending.dm:
```diff
@@ -1587,6 +1618,385 @@
+			productListUpdater(targetContainer, user)
+			user.visible_message("<b>[user.name]</b> loads [target] into [src].")
+			return
+		else if (!action || (usr.stat || usr.restrained() || !in_interact_range(src, user)))
```
ditto

ZeWaka commented on code/modules/vending/vending.dm:
```diff
@@ -1587,6 +1618,385 @@
+		if (status & (BROKEN|NOPOWER))
+			setCrtOverlayStatus(0)
+			return
+		if (usr.stat || usr.restrained() || !in_interact_range(src, usr))
```
ditto

ZeWaka commented on code/modules/vending/vending.dm:
```diff
@@ -1587,6 +1618,385 @@
+		else if (href_list["rename"] && src.panel_open && src.unlocked)
+			var/inp
+			inp = html_encode(sanitize(input(usr,"Enter new name:","Vendor Name", "") as text))
+			if(inp && inp != "" && (usr.stat || usr.restrained() || in_interact_range(src, usr)))
```
ditto

ZeWaka commented on code/modules/vending/vending.dm:
```diff
@@ -1587,6 +1618,385 @@
+		else if (href_list["setprice"] && src.panel_open && src.unlocked)
+			var/inp
+			inp = input(usr,"Enter the new price:","Item Price", "") as num
+			if(inp && inp >= 0 && (usr.stat || usr.restrained() || in_interact_range(src, usr)))
```
ditto

penny64 commented:
Alright I use the correct interaction checks now, I fixed the promo image bug, and I fixed all appearance bugs with an appearance manager that I use everywhere, including as an override for the power state change proc!

Azrun commented on code/modules/vending/vending.dm:
```diff
@@ -1587,6 +1618,414 @@
+				var/datum/data/vending_product/player_product/R = locate(href_list["setprice"]) in src.player_list
+				R.product_cost = inp
+				src.generate_HTML(1, 0)
+		else if (href_list["vend"] && (length(player_list) <= 0))
```
!length(foo)

penny64 commented on code/modules/vending/vending.dm:
```diff
@@ -1587,6 +1618,414 @@
+		wiresdesc = "[desc] Nothing has been wired up."
+		glassdesc = "[desc] Isn't there usually glass?"
+		readydesc = "[desc] Just needs a few screws tightened."
+	proc/setFrameState(state, mob/user, obj/item/target)
```
Else where? This is just some static strings I setup

penny64 replied:
Ah sure I'll check it out!

I think I have all of your stuff fixed now, just have to test and push when I get back home

ZeWaka commented on code/modules/vending/vending.dm:
```diff
@@ -1753,7 +1756,10 @@
+				qdel(src)
+		else
+			setFrameState("UNWRENCHED")
+			CRASH("Invalid state \"[state]\" set in [src] construction process")
```
i would notate this better as to where the runtime is coming from (x,y,z,[src])

Azrun reviewed (approved):
12 cases of `if(`
79 instances of `if (`

Consistent style within same function/file (😆)/codebase (🤣) improves readability.

penny64 commented:
> 12 cases of `if(`
> 79 instances of `if (`
> 
> Consistent style within same function/file (😆)/codebase (🤣) improves readability.

I'm actually sitting down with it right now, there's some == too!

penny64 commented:
@ZeWaka don't think I'm getting any more reviews, whenever you want to look at this :>

ZeWaka commented on code/datums/abilities/wraith.dm:
```diff
@@ -243,7 +243,7 @@
 			. = W.makeRevenant(T)		//return 0
 			if(!.)
-				playsound(W.loc, "sound/voice/wraith/reventer.ogg", 86, 0)
+				playsound(W.loc, "sound/voice/wraith/reventer.ogg", 80, 0)
```
What's with all these volume adjustments?

ZeWaka commented on code/modules/vending/vending.dm:
```diff
@@ -58,7 +59,7 @@
 	var/list/product_list = new() //List of datum/data/vending_product
 	var/glitchy_slogans = 0 // do they come out aLL FunKY lIKe THIs?
-
+	var/player_list // For the player vending machines
```
```suggestion
	/// For player vending machines
	var/player_list 
```
Doc comments

ZeWaka reviewed (approved):
Looks good otherwise
